### PR TITLE
rgw/multisite-notification: retry storing bucket notification attrs for ECANCELED(ConcurrentModification) errors.

### PR DESCRIPTION
**Pull request by kchheda3**
Fixes: https://tracker.ceph.com/issues/65590

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [x] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/rgw/rgw_op.h [hunk 1]
```diff
@@ -165,6 +165,37 @@ int rgw_rest_get_json_input(CephContext *cct, req_state *s, T& out,
...
+// again. We have, however, to try the right *part* again.  We can't
+// simply re-send, since that will obliterate the previous update.
+//
+// Thus, callers of this function should include everything that
+// merges information to be changed into the bucket information as
+// well as the call to set it.
+//
+// The called function must return an integer, negative on error. In
+// general, they should just return op_ret.
+namespace {
```
Comment (cbodley) [inline src/rgw/rgw_op.h:183]: this was in the anonymous namespace to make it static to rgw_op.cc. please remove the namespace block

**Review by cbodley — COMMENTED** [review]
👍

**Comment by cbodley** [comment]
did some local testing on this. it took a while to reproduce a ECANCELED error in RGWPSCreateNotifOp (they mostly happened in the metadata sync thread). eventually i bumped the number of zones to 8 and caught one:
```
2024-04-19T12:20:40.887-0400 7fec3a2006c0  1 -- 192.168.245.130:0/1590865445 <== osd.0 v2:192.168.245.130:6826/2794811376 8106 ==== osd_op_reply(8050 .bucket.meta.jhajal-1:9039f7f8-6dd6-4737-8cfa-1df8c370413a.4176.1 [delete,create,call,call,writefull 0~382,setxattr (185),setxattr (457)] v16'32 uv31 ondisk = -125 ((125) Operation canceled)) ==== 461+0+0 (crc 0 0 0) 0x55eab6eed440 con 0x55eab2c24880
2024-04-19T12:20:40.887-0400 7feba92006c0 10 req 13273874293440243183 0.017000554s s3:pubsub_notification_create_s3 removing a4.rgw.meta+root+.bucket.meta.jhajal-1:9039f7f8-6dd6-4737-8cfa-1df8c370413a.4176.1 from cache
```
we try again and succeed:
```
2024-04-19T12:20:40.890-0400 7febad8006c0  1 -- 192.168.245.130:0/1590865445 --> [v2:192.168.245.130:6826/2794811376,v1:192.168.245.130:6827/2794811376] -- osd_op(unknown.0.0:8054 4.0 4:62c1eabf:root::.bucket.meta.jhajal-1%3a9039f7f8-6dd6-4737-8cfa-1df8c370413a.4176.1:head [delete,create,call version.check_conds in=122b,call version.set in=58b,writefull 0~382 in=382b,setxattr user.rgw.acl (185) in=197b,setxattr user.rgw.bucket-notification (1269) in=1297b] snapc 0=[] ondisk+write+known_if_redirected+supports_pool_eio e16) -- 0x55eab2bd1500 con 0x55eab2c24880
2024-04-19T12:20:40.892-0400 7fec3a2006c0  1 -- 192.168.245.130:0/1590865445 <== osd.0 v2:192.168.245.130:6826/2794811376 8110 ==== osd_op_reply(8054 .bucket.meta.jhajal-1:9039f7f8-6dd6-4737-8cfa-1df8c370413a.4176.1 [delete,create,call,call,writefull 0~382,setxattr (185),setxattr (1269)] v16'33 uv33 ondisk = 0) ==== 461+0+0 (crc 0 0 0) 0x55eab27b0b40 con 0x55eab2c24880
```
and the op returns success 👍 
```
2024-04-19T12:20:40.899-0400 7febaa6006c0  2 req 13273874293440243183 0.029000944s s3:pubsub_notification_create_s3 http status=200
2024-04-19T12:20:40.899-0400 7febaa6006c0  1 ====== req done req=0x7fec3f3fe650 op status=0 http_status=200 latency=0.029000944s ======
```

File: src/rgw/rgw_rest_pubsub.cc [hunk 2]
```diff
@@ -1318,48 +1318,49 @@ void RGWPSCreateNotifOp::execute_v2(optional_yield y) {
...
+      return remove_notification_v2(this, driver, s->bucket.get(),
+                                    /*delete all notif=true*/"", y);
     }
-    const auto& topic_name = arn->resource;
-
-    auto t = topics.find(*arn);
-    if (t == topics.end()) {
-      continue;
+    rgw_pubsub_bucket_topics bucket_topics;
+    op_ret = get_bucket_notifications(this, s->bucket.get(), bucket_topics);
```
Comment (yuvalif) [inline src/rgw/rgw_rest_pubsub.cc:1327]: thtat's a bit confusing. probably better to use a local "ret" var here, and not change the `op_ret` which is passed by reference.

File: src/rgw/rgw_rest_pubsub.cc [hunk 3]
```diff
@@ -1318,48 +1318,49 @@ void RGWPSCreateNotifOp::execute_v2(optional_yield y) {
...
-        << s->bucket << ", ret = " << op_ret << dendl;
-    return;
-  }
-  for (const auto& c : configurations.list) {
-    const auto& notif_name = c.id;
-
-    const auto arn = rgw::ARN::parse(c.topic_arn);
-    if (!arn) { // already validated above
-      continue;
+  op_ret = retry_raced_bucket_write(this, s->bucket.get(), [&] {
```
Comment (yuvalif) [inline src/rgw/rgw_rest_pubsub.cc:1321]: what capture everything by ref?
what do yo uactually need?

**Comment by cbodley** [comment]
passed qa in https://pulpito.ceph.com/cbodley-2024-04-24_12:59:55-rgw-wip-cbodley-testing-distro-default-smithi/ with rerun https://pulpito.ceph.com/cbodley-2024-04-24_20:45:10-rgw-wip-cbodley-testing-distro-default-smithi/

**Comment by cbodley** [comment]
just waiting for review comments to be resolved here

File: src/rgw/rgw_rest_pubsub.cc (hunk 3 above)
Comment (kchheda3) [reply src/rgw/rgw_rest_pubsub.cc:1321]: > what capture everything by ref? what do yo uactually need?

it was just lazy coding, did not  bother checking the params needed
few places were just doing same of capturing all, so i did same.
will it matter ? or change and then we have to re run tuethology ?

**Comment by kchheda3** [comment]
> just waiting for review comments to be resolved here

sorry i had missed the comment

File: src/rgw/rgw_rest_pubsub.cc (hunk 3 above)
Comment (yuvalif) [reply src/rgw/rgw_rest_pubsub.cc:1321]: > will it matter

IMO, it mainly matters for the reader of the code, who needs to figure out what this lambda might change and/or need. but, to me, this is an important reason to be explicit as much as possible with capturing

File: src/rgw/rgw_rest_pubsub.cc (hunk 2 above)
Comment (kchheda3) [reply src/rgw/rgw_rest_pubsub.cc]: done

File: src/rgw/rgw_rest_pubsub.cc (hunk 3 above)
Comment (kchheda3) [reply src/rgw/rgw_rest_pubsub.cc]: done

**Comment by cbodley** [comment]
> No space left on device

**Comment by cbodley** [comment]
jenkins test submodules

**Pull request merged**